README.md
# Anobii2Goodreads

This script converts a standard CSV export from the Anobii website to a format you should be able to import into Goodreads.

- Anobii homepage: [https://anobii.com](https://anobii.com)
- Goodreads homepage: [https://goodreads.com](https://goodreads.com)

# How to export from Anobii and import to Goodreads

## Prerequisites

Python v2.7.x installed

The script contains a configuration section to manage different language translations. Italian is the default one; if you need to manage a different language, you must customize the following constants basing on your Anobii export content file:

|||
|------|-------|
|`FINISHED`|Book you finished|
|`DROPPED`|Book you gave up on|
|`READING`|Book you're currently reading|
|`MONTHS`|List of months|

## Export from Anobii

1. Login to Anobii and access your books library page
2. Click on the option button you can find on the top right, besides ordering filters
3. Select "Export list" option and choose CSV format 
4. You will receive an email in a few minutes with a CSV file attached. Download this file and rename it as `anobii_export.csv`.

## Convert from Anobii to Goodreads

1. To convert between the export format of Anobii and import format of Goodreads, we use the [Anobii2Goodreads](https://github.com/marcosegato/Anobii2Goodreads) script. Git clone it or download the code:
    ```shell
    $ git clone https://github.com/marcosegato/Anobii2Goodreads
    ```
2. Place the `anobii_export.csv` file in the same directory containing the `anobii-to-goodreads.py` script.
3. Perform the conversion, the output is written to `import_to_goodreads.csv` file:
    ```shell
    $ python anobii-to-goodreads.py
    ```
## Import to Goodreads

1. Login to Goodreads. Go to its [import page](https://goodreads.com/review/import/) and import the `import_to_goodreads.csv` file.
2. Refresh this page to see the import progress. Few books may not be imported (e.g. missing ISBN), consider [adding them manually](https://www.goodreads.com/book/new).
3. Go to your own update stream, you should be able to see all the imports. Do not worry, Goodreads only shows a few of these updates on your friends' stream.

# Credits

This script is based on the following works:

* Original version by Giacomo Lacava published on 2008: 
[http://blog.pythonaro.com/2008/08/export-your-books-from-anobii-to_14.html](http://blog.pythonaro.com/2008/08/export-your-books-from-anobii-to_14.html)
* Improvements by Tijs Teulings to make it work again with the latest export format on 2015: [https://codeyarns.com/2015/07/03/how-to-export-from-anobii-and-import-to-goodreads/](https://codeyarns.com/2015/07/03/how-to-export-from-anobii-and-import-to-goodreads/)

# License
GNU General Public License version 3

CHANGELOG.md
# v1.1.0
##  19/04/2020

    * New configuration section to manage different language translations (Italian as default)
    * Exclude books with invalid ISBN

# v1.0.0
##  18/04/2020

    * Fix title import
    * Add publication year import
    * Add reading status import
    * Add read date import
    * Removed tag import in bookshelves
    * Project documentation improved

anobii-to-goodreads.py
# Customize these variables to define input and output
anobii_file = "anobii_export.csv"
goodreads_file = "import_to_goodreads.csv" 

# Customize language translations
FINISHED = "Finito nel "
DROPPED  = "Abbandonato nel "
READING  = "In lettura dal "
MONTHS   = ["gen", "feb", "mar", "apr", "mag", "giu", "lug", "ago", "set", "ott", "nov", "dic"]

####### do not change anything below this line

from datetime import date
import csv, codecs, cStringIO

class UTF8Recoder:
	"""
	Iterator that reads an encoded stream and reencodes the input to UTF-8
	"""
	def __init__(self, f, encoding):
		self.reader = codecs.getreader(encoding)(f)
	
	def __iter__(self):
		return self
	
	def next(self):
		return self.reader.next().encode("utf-8")

class UnicodeReader:
	"""
	A CSV reader which will iterate over lines in the CSV file "f",
	which is encoded in the given encoding.
	"""
	
	def __init__(self, f, dialect=csv.excel, encoding="utf-8", **kwds):
		f = UTF8Recoder(f, encoding)
		self.reader = csv.reader(f, dialect=dialect, **kwds)
	
	def next(self):
		row = self.reader.next()
		return [unicode(s, "utf-8") for s in row]
	
	def __iter__(self):
		return self

class UnicodeWriter:
	"""
	A CSV writer which will write rows to CSV file "f",
	which is encoded in the given encoding.
	"""
	
	def __init__(self, f, dialect=csv.excel, encoding="utf-8", **kwds):
		# Redirect output to a queue
		self.queue = cStringIO.StringIO()
		self.writer = csv.writer(self.queue, dialect=dialect, **kwds)
		self.stream = f
		self.encoder = codecs.getincrementalencoder(encoding)()

	def writerow(self, row):
		items = []
		for s in row:
			if type(s) == type(u"s"):
				items.append(s.encode("utf8"))
			else:
				items.append(s)

		self.writer.writerow(items)
		# Fetch UTF-8 output from the queue ...
		data = self.queue.getvalue()
		data = data.decode("utf-8")
		# ... and reencode it into the target encoding
		data = self.encoder.encode(data)
		# write to the target stream
		self.stream.write(data)
		# empty queue
		self.queue.truncate(0)

	def writerows(self, rows):
		for row in rows:
			self.writerow(row)


reader = UnicodeReader(open(anobii_file,"rb"))
reader.next() # first line is column titles
target = []
target.append(["Title","Author","Additional Authors","ISBN","ISBN13","My Rating","Average Rating","Publisher","Binding","Year Published","Original Publication Year","Date Read","Date Added","Bookshelves","My Review","Spoiler","Private Notes","Recommended For","Recommended By"])
# loading all in memory is not efficient, there's certainly a better way
for l in reader:
	# isbn
	isbn = l[0].replace("'","")
	if isbn is None or isbn == "":
		print "Invalid ISBN in '" + l[1] + "' book, consider adding it manually"
		continue
	# title
	if l[2] == "":
		title = l[1]
	else:
		title = l[1] + ". " + l[2]
	# author
	author = l[3]
	# binding
	binding = l[4]
	# pages
	pages = l[5]
	# publisher
	publisher = l[6]
	# pubdate
	pubdate = (l[7])[1:5]
	# privnote
	privnote = l[8]
	# comment
	comment = l[10]
	# status
	status = l[11]
	# bookshelves
	bookshelves = "to-read";
	if status.startswith(FINISHED): bookshelves = "read"
	if status.startswith(READING):  bookshelves = "currently-reading"
	if status.startswith(DROPPED):  bookshelves = "gave-up-on"
	# readdate
	tmpreaddate = l[11].replace(", 00:00:00","")
	yreaddate = tmpreaddate[-4:]
	mtmpreaddate = tmpreaddate.replace(yreaddate,"")[-5:]
	mreaddate = ""
	if mtmpreaddate.strip() == MONTHS[0]:  mreaddate = "01"
	if mtmpreaddate.strip() == MONTHS[1]:  mreaddate = "02"
	if mtmpreaddate.strip() == MONTHS[2]:  mreaddate = "03"
	if mtmpreaddate.strip() == MONTHS[3]:  mreaddate = "04"
	if mtmpreaddate.strip() == MONTHS[4]:  mreaddate = "05"
	if mtmpreaddate.strip() == MONTHS[5]:  mreaddate = "06"
	if mtmpreaddate.strip() == MONTHS[6]:  mreaddate = "07"
	if mtmpreaddate.strip() == MONTHS[7]:  mreaddate = "08"
	if mtmpreaddate.strip() == MONTHS[8]:  mreaddate = "09"
	if mtmpreaddate.strip() == MONTHS[9]:  mreaddate = "10"
	if mtmpreaddate.strip() == MONTHS[10]: mreaddate = "11"
	if mtmpreaddate.strip() == MONTHS[11]: mreaddate = "12"
	dtmpreaddate = tmpreaddate.replace(yreaddate,"").replace(mtmpreaddate,"")[-2:]
	dreaddate = dtmpreaddate.replace(" ","0")
	readdate = yreaddate + "-" + mreaddate + "-" + dreaddate
	if readdate == "1970-01-01": readdate = ""
	if readdate == "--": readdate = ""
	# dateadded
	dateadded = readdate
	# recover readdate basing on bookshelves
	if bookshelves == "currently-reading": readdate = ""
	if bookshelves == "gave-up-on": readdate = ""
	# rating
	rating = l[12]
	
	tline = [title,author,"",isbn,"",rating,"",publisher,binding,pubdate,"",readdate,dateadded,bookshelves,comment,"",privnote,"",""]
	target.append(tline)

writer = UnicodeWriter(open(goodreads_file,"wb"),dialect='excel',quoting=csv.QUOTE_NONNUMERIC)
writer.writerows(target)

print "Done! saved output to " + goodreads_file